WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift

```diff
@@ -295,10 +308,17 @@ private extension PointOfSaleAggregateModel {
             .assign(to: &$cardPresentPaymentInlineMessage)
 
         cardPresentPaymentService.paymentEventPublisher
-            .compactMap { [weak self] paymentEvent in
-                guard let self else { return .none }
-                return PointOfSalePaymentState(from: paymentEvent,
-                                               using: presentationStyleDeterminerDependencies)
+            .compactMap { [weak self] paymentEvent -> PointOfSalePaymentState? in
+                guard let self else { return nil }
+
+                let newPaymentState = PointOfSalePaymentState(from: paymentEvent,
+                                                              using: presentationStyleDeterminerDependencies)
+
+                if self.isCashPaymentInProgress, case .card = newPaymentState {
+                    cardPresentPaymentService.cancelPayment()
+                    return .cash(.paymentSuccess)
```

Comment:
Overriding the state here with `.cash(.paymentSuccess)` doesn't feel quite right and too optimistic, but since this only can happen when we're already showing the "success" view, I don't see much problem returning it 🤔 Combining this filter with the `isCashPaymentInProgress` flag is the only way I could find to intercept card events and cancel the payment attempt.

Reply:
Soo... this exposed an issue with duplicating state (`isCashPaymentInProgress`). 

When testing, the first time that I tapped a card at the same time as tapping `mark payment complete`, I got a blank screen after seeing the success flash up.

I exited POS (my only option) and came back in, but `isCashPaymentInProgress` was still true then, and so every attempt to take a card payment after that failed because it was immediately cancelled (I think.)

Not something to go and separately fix – this is a symptom of issues with the approach of keeping separate flags for what's going on in the aggregate model; they can get out of sync and cause strange bugs.

Reply:
Yeah, it's more than too optimistic, it's flawed logic, as it shows the success screen even if payment hasn't happened:

1. go to checkout using an error amount
2. connect the reader
3. tap collect cash payment but don't enter anything
4. tap a card on the reader

You'll see the cash success screen (which is incorrect anyway), and then it'll go in to card payment, which will fail because it gets cancelled. Then it goes to a blank screen which you can't navigate away from without exiting POS. The order doesn't get marked as paid, but the merchant has seen a success screen.

This can also be reproduced using a non-error amount, if the first card tap doesn't register correctly – if you kind of bring it in towards the reader slowly from far away, so it asks you to tap again.